### PR TITLE
Fixed Point Numbers & Operations

### PR DESCRIPTION
**Pull request by chrisnevers**
Add fixed point numbers and almost all operations needed for `amm`. Still having some issues working out `pow` for fixed point numbers. I did add a `pow` function for regular `UInt`s though.

### CODE REVIEW

File: docs-src/ref-programs.scrbl [hunk 1]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+  export const FixedPoint = Object({ scale: UInt, i: UInt }); }
+
+@reachin{FixedPoint} can be used to represent numbers with a fixed number of digits after the decimal point.
+They are handy for representing fractional values, especially in base 10. The value of a fixed point number is determined
+by dividing the underlying integer value, @tt{i}, by its scale factor, @tt{scale}. For example, we could
+represent the value @reachin{1.234} with @reachin{{ scale: 1000, i : 1234 }} or @reachin{mk_fixed_point(1000)(1234)}.
+A scale factor of @tt{1000} correlates to 3 decimal places of precision. Similarly, a scale factor of @tt{100} would
+have 2 decimal places of precision.
+
+@(mint-define! '("mk_fixed_point"))
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1918]: Let's abbreviate fixpoints as `fx` and just call this function `fx`

File: docs-src/ref-programs.scrbl [hunk 2]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+@reach{
+  const scale = 10;
+  const i = 56;
+  mk_fixed_point(scale)(i) // represents 5.6 }
+
+@index{mk_fixed_point} @reachin{mk_fixed_point(scale)(i)} will return a function that can be used to
+instantiate fixed point numbers with a particular scale factor.
+
+@(mint-define! '("fixed_point_convert_scale"))
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1927]: `fxscale` or `fxrescale`?

File: docs-src/ref-programs.scrbl [hunk 3]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+@reach{
+  const x = mk_fixed_point(1000)(1234); // x = 1.234
+  fixed_point_convert_scale(x, 100);    // => 1.23 }
+
+@index{fixed_point_convert_scale} @reachin{fixed_point_convert_scale(x, scale)} will convert a fixed point number from using
+one scale to another. This operation can result in loss of precision, as demonstrated in the above example.
+
+@(mint-define! '("fixed_point_homogenize_scales"))
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1935]: This is a descriptive name and I hate it. I want to say `fxmeet` but that's a bad (too technical name), please think of something else :)

File: docs-src/ref-programs.scrbl [hunk 4]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+@reach{
+  const x = mk_fixed_point(1000)(824345); // x = 824.345
+  const y = mk_fixed_point(100)(4567);    // y =  45.67
+  fixed_point_homogenize_scales(x, y);    // => [ 1000, 824.345, 45.670 ] }
+
+@index{fixed_point_homogenize_scales} @reachin{fixed_point_homogenize_scales(x, y)} will convert the fixed point numbers
+to use the same scale. The larger scale of the two arguments will be chosen. The function will return a @tt{3-tuple} consisting
+of the common scale and the newly scaled values.
+
+@index{fixed_point_add} @reachin{fixed_point_add(x, y)} adds two fixed point numbers.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1945]: `fxadd`

File: docs-src/ref-programs.scrbl [hunk 5]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+
+@index{fixed_point_sub} @reachin{fixed_point_sub(x, y)} subtracts two fixed point numbers.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1947]: `fxsub`

File: docs-src/ref-programs.scrbl [hunk 6]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+
+@index{fixed_point_mul} @reachin{fixed_point_mul(x, y)} multiplies two fixed point numbers.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1949]: `fxmul`

File: docs-src/ref-programs.scrbl [hunk 7]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+
+@index{fixed_point_div} @reachin{fixed_point_div(x, y, scale_factor)} divides two fixed point numbers. The numerator, @tt{x},
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1951]: `fxdiv`

File: docs-src/ref-programs.scrbl [hunk 8]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+will be multiplied by the scale factor to provide a more precise answer. For example,
+
+@reach{
+  fixed_point_div(34.56, 1.234, 10)     // => 28
+  fixed_point_div(34.56, 1.234, 100000) // => 28.0064 }
+
+@index{fixed_point_sqrt} @reachin{fixed_point_sqrt(x, k)} approximates the sqrt of the fixed number, @tt{x}, using
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1958]: `fxsqrt`

File: docs-src/ref-programs.scrbl [hunk 9]
```diff
@@ -1890,3 +1890,73 @@ of iterations the algorithm should perform.
...
+@tt{k} iterations of the @reachin{sqrt} algorithm.
+
+@index{fixed_point_cmp} @reachin{fixed_point_cmp(op, x, y)} applies the comparison
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1961]: `fxcmp`

Comment (jeapostrophe) [reply docs-src/ref-programs.scrbl:1961]: Potentially pre-define `fxlt`, `fxle`, `fxeq`, and so on

File: hs/rsh/stdlib.rsh [hunk 10]
```diff
@@ -203,3 +203,70 @@ export const sqrt = (y, k) =>
...
+export const fixed_point_homogenize_scales = (x, y) => {
+  const scale = x.scale < y.scale ? y.scale : x.scale;
+  const x_ = fixed_point_convert_scale(x, scale);
+  const y_ = fixed_point_convert_scale(y, scale);
+  return [ scale, x_, y_ ];
+}
+
+export const fixed_point_add = (x, y) => {
+  const [ scale, x_, y_ ] = fixed_point_homogenize_scales(x, y);
+  assert(x_.i + y_.i <= UInt.max, "fixed_point_add: Addition causes overflow.");
```
Comment (jeapostrophe) [inline hs/rsh/stdlib.rsh:229]: Why is this needed... given that `verifyOverflow` will add this when necessary?

File: hs/rsh/stdlib.rsh [hunk 11]
```diff
@@ -203,3 +203,70 @@ export const sqrt = (y, k) =>
...
+  return { i: x_.i + y_.i, scale };
+}
+
+export const fixed_point_sub = (x, y) => {
+  const [ scale, x_, y_ ] = fixed_point_homogenize_scales(x, y);
+  assert(x_.i - y_.i >= 0, "fixed_point_sub: Subtraction causes underflow.");
```
Comment (jeapostrophe) [inline hs/rsh/stdlib.rsh:235]: Ditto, but also this expression causes the overflow to happen

File: hs/rsh/stdlib.rsh [hunk 12]
```diff
@@ -203,3 +203,70 @@ export const sqrt = (y, k) =>
...
+  return { i: x_.i - y_.i, scale };
+}
+
+export const fixed_point_mul = (x, y) => {
+  assert(x.scale * y.scale <= UInt.max,
+    "fixed_point_mul: Multiplication of scales overflow.");
+  assert(x.i * y.i <= UInt.max,
```
Comment (jeapostrophe) [inline hs/rsh/stdlib.rsh:242]: Ditto

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

**Comment by jeapostrophe** [comment]
Why does the "obvious thing" for `fxpow` not work? Can I help in some way?

File: docs-src/ref-programs.scrbl (hunk 3 above)
Comment (chrisnevers) [reply docs-src/ref-programs.scrbl:1935]: This one took me a while to find a suitable name 😆 I'll rename

File: hs/rsh/stdlib.rsh (hunk 10 above)
Comment (chrisnevers) [reply hs/rsh/stdlib.rsh:229]: Oh duh, I'll remove these `assert`s

**Comment by chrisnevers** [comment]
@jeapostrophe I think the assertions in `fxmul` were given me difficulties with implementing `fxpow`. Now that I removed them, it seems to work ok

File: docs-src/ref-programs.scrbl [hunk 13]
```diff
@@ -1911,52 +1911,72 @@ point of overflow.
...
+@index{fxge} @reachin{fxge(x, y)} tests whether @tt{x} is greater than or equal to @tt{y}.
+
+@index{fxeq} @reachin{fxeq(x, y)} tests whether @tt{x} is equal to @tt{y}.
+
+@index{fxne} @reachin{fxne(x, y)} tests whether @tt{x} is not equal to @tt{y}.
+
+@index{fxpowi} @reachin{fxpowi(base, power, precision)} approximates the power of the fixed number, @tt{base},
+raised to the @reachin{UInt}, @tt{power}. The third argument must be an @reachin{UInt} whose value is known
+at compile time, which represents the number of iterations the algorithm should perform. For reference, @tt{10}
+iterations provides accuracy until the point of overflow.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1981]: I realized that this comment about "overflow" implies something about what you think `UInt.max` is

File: docs-src/ref-programs.scrbl [hunk 14]
```diff
@@ -1899,8 +1899,7 @@ square root up to @tt{580} squared, or @tt{336,400}.
 
 @index{pow} @reachin{pow(base, power, precision)} Calculates the approximate value of raising base to power.
 The third argument must be an @reachin{UInt} whose value is known at compile time, which represents the number
-of iterations the algorithm should perform. For reference, @tt{10} iterations provides accuracy until the
-point of overflow.
+of iterations the algorithm should perform.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1902]: I wasn't saying to remove it entirely... just to add more context and give examples.

File: docs-src/ref-programs.scrbl [hunk 15]
```diff
@@ -1955,9 +1954,22 @@ will be multiplied by the scale factor to provide a more precise answer. For exa
   fxdiv(34.56, 1.234, 10)     // => 28
   fxdiv(34.56, 1.234, 100000) // => 28.0064 }
 
+@index{fxmod} @reachin{fxmod(x, y)} finds the remainder of dividing @tt{x} by @tt{y}.
+
+@index{fxfloor} @reachin{fxfloor(x)} rounds the fixed point number, @tt{x}, down to the nearest whole number.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1959]: Does it return another `fx` or does it return a `UInt`?

File: docs-src/ref-programs.scrbl [hunk 16]
```diff
@@ -1955,9 +1954,22 @@ will be multiplied by the scale factor to provide a more precise answer. For exa
...
+
 @index{fxsqrt} @reachin{fxsqrt(x, k)} approximates the sqrt of the fixed number, @tt{x}, using
 @tt{k} iterations of the @reachin{sqrt} algorithm.
 
+@index{fxpow} @reachin{fxpow(base, power, precision, scalePrecision)} approximates the power of the fixed number, @tt{base},
+raised to the fixed point number, @tt{power}. The third argument must be an @reachin{UInt} whose value is known
+at compile time, which represents the number of iterations the algorithm should perform. The @tt{scalePrecision} argument must
+be a @tt{UInt} and represents the scale of the return value.
```
Comment (jeapostrophe) [inline docs-src/ref-programs.scrbl:1967]: These need some advice about precision

File: docs-src/ref-programs.scrbl (hunk 15 above)
Comment (jeapostrophe) [reply docs-src/ref-programs.scrbl:1959]: Yea, you should specify it
